test/layout-test.ts

```diff
@@ -119,6 +251,64 @@ describe('layout', () => {
       assert.deepEqual(breakpoints, [0]);
     });
 
+    it('generates narrow frog prince layout from p. 81 of Digital Typography', () => {
+      const items = frogPrinceItems();
+      // width given on p. 78 of Digital Typography
+      // subtract 1em (18 machine units) from the first line
+      const lineLengths = [372, ...Array(items.length - 1).fill(390)];
+      const breakpoints = breakLines(items, lineLengths);
+      const lines = lineStrings(items, breakpoints);
+      assert.deepEqual(lines, [
+        'In olden times when wishing still helped one,',
+        'there lived a king whose daughters were all beau-',
+        'tiful; and the youngest was so beautiful that the',
+        'sun itself, which has seen so much, was aston-',
+        'ished whenever it shone in her face. Close by the',
+        'king’s castle lay a great dark forest, and under an',
+        'old limetree in the forest was a well, and when',
+        'the day was very warm, the king’s child went out',
+        'into the forest and sat down by the side of the',
+        'cool fountain; and when she was bored she took a',
+        'golden ball, and threw it up on high and caught',
+        'it; and this ball was her favorite plaything. -',
+      ]);
+      const adjRatios = adjustmentRatios(items, lineLengths, breakpoints).map((num) =>
+        Number(num.toFixed(3)),
+      );
+      assert.deepEqual(
+        adjRatios,
+        [0.857, 0.0, 0.28, 1.0, 0.067, -0.278, 0.536, -0.167, 0.7, -0.176, 0.357, 0.049],
+      );
+    });
+
+    it('generates wide frog prince layout from p. 82 of Digital Typography', () => {
+      const items = frogPrinceItems();
+      // width given on p. 81 of Digital Typography
+      // subtract 1em (18 machine units) from the first line
+      const lineLengths = [482, ...Array(items.length - 1).fill(500)];
+      const breakpoints = breakLines(items, lineLengths);
+      const lines = lineStrings(items, breakpoints);
+      assert.deepEqual(lines, [
+        'In olden times when wishing still helped one, there lived a',
+        'king whose daughters were all beautiful; and the youngest was',
+        'so beautiful that the sun itself, which has seen so much, was',
+        'astonished whenever it shone in her face. Close by the king’s',
+        'castle lay a great dark forest, and under an old limetree in the',
+        'forest was a well, and when the day was very warm, the king’s',
+        'child went out into the forest and sat down by the side of the',
+        'cool fountain; and when she was bored she took a golden ball,',
+        'and threw it up on high and caught it; and this ball was her',
+        'favorite plaything. -',
+      ]);
+      const adjRatios = adjustmentRatios(items, lineLengths, breakpoints).map((num) =>
+        Number(num.toFixed(3)),
+      );
+      assert.deepEqual(
+        adjRatios,
+        [0.774, 0.179, 0.629, 0.545, 0.0, 0.079, 0.282, 0.294, 0.575, 0.353],
```

Comment:
The last adjustment ratio value also differs from the book here.

Reply:
Ditto